src/imageattributes.c

```diff
@@ -204,23 +218,100 @@ gdip_process_bitmap_attributes (GpBitmap *bitmap, void **dest, GpImageAttributes
 
 	}
 
+	/* Treshold correction */
+	if (!(treshold->flags & ImageAttributeFlagsNoOp) && treshold->flags & ImageAttributeFlagsThresholdEnabled) {
+		BYTE cutoff = (BYTE)round(treshold->threshold * 255.0);
+
+		for (int y = 0; y < bitmap->active_bitmap->height; y++) {
+			for (int x = 0; x < bitmap->active_bitmap->width; x++) {
+
+				BYTE r, g, b, a;
+
+				GdipBitmapGetPixel(bmpdest, x, y, &color);
+
+				a = (color & 0xff000000) >> 24;
+				r = (color & 0x00ff0000) >> 16;
+				g = (color & 0x0000ff00) >> 8;
+				b = (color & 0x000000ff);
+
+				r = r > cutoff ? 255 : 0;
+				g = g > cutoff ? 255 : 0;
+				b = b > cutoff ? 255 : 0;
+
+				color = ((guint32)a << 24) | (r << 16) | (g << 8) | b;
+				GdipBitmapSetPixel(bmpdest, x, y, color);
+			}
+		}
+	}
+
+	/* CMYK calculation */
+	if (!(cmyk->flags & ImageAttributeFlagsNoOp) && cmyk->flags & ImageAttributeFlagsOutputChannelEnabled) {
+		for (int y = 0; y < bitmap->active_bitmap->height; y++) {
+			for (int x = 0; x < bitmap->active_bitmap->width; x++) {
+
+				BYTE r, g, b, a, C, M, Y, K;
+
+				GdipBitmapGetPixel(bmpdest, x, y, &color);
+
+				a = (color & 0xff000000) >> 24;
+				r = (color & 0x00ff0000) >> 16;
+				g = (color & 0x0000ff00) >> 8;
+				b = (color & 0x000000ff);
+
+				C = 255 - r;
+				M = 255 - g;
+				Y = 255 - b;
+				K = min(min(C, M), Y);
+				
+				/* correct complementary color lever based on k */
+				C -= K;
+				M -= K;
+				Y -= K;
+
+				switch (cmyk->outputchannel_flags) {
+				case ColorChannelFlagsC:
+					r = g = b = C;
+					break;
+				case ColorChannelFlagsM:
+					r = g = b = M;
+					break;
+				case ColorChannelFlagsY:
+					r = g = b = Y;
+					break;
+				case ColorChannelFlagsK:
+					r = g = b = K;
+					break;
+				default:
+					return InvalidParameter;
+				}
+
+				color = ((guint32)a << 24) | (r << 16) | (g << 8) | b;
+				GdipBitmapSetPixel(bmpdest, x, y, color);
+			}
+		}
+	}
+
 	/* Apply transparency range */
-	if (trans->flags & ImageAttributeFlagsColorKeysEnabled) {
+	/* FIXME: This will convert pixels which have a color in [key_colorlow, key_colorhey] to transparent pixels.
+	          However, GdipDrawImageRectRect in image.c uses cairo_fill to draw these transparent pixels onto
+			  the target surface. This method will not copy the transparency value; rather, it will ignore
+			  transparent values. */
+	if (!(trans->flags & ImageAttributeFlagsNoOp) && trans->flags & ImageAttributeFlagsColorKeysEnabled) {
 		for (int y = 0; y < bitmap->active_bitmap->height; y++) {
 			for (int x = 0; x < bitmap->active_bitmap->width; x++) {
 
 				GdipBitmapGetPixel (bmpdest, x, y, &color);
-
+				color &= ~ALPHA_MASK;
```

Comment:
@qmfrederik Do you remember what was the rationale for this changed line?

It breaks some user code (https://github.com/mono/mono/issues/8514) because it masks only the input value and not the values in `trans->key_colorlow` and `trans->key_colorhigh`. Thus the condition for transparent pixels is not matched. It should either mask on both places or neither of them.

Reply:
Wine compares just the R, G, B values do I'd guess that the correct way is to mask both.